plugins/timertop/timerinfo.h

```diff
@@ -28,67 +28,85 @@
 #ifndef GAMMARAY_TIMERTOP_TIMERINFO_H
 #define GAMMARAY_TIMERTOP_TIMERINFO_H
 
-#include "functioncalltimer.h"
-
-#include <QSharedPointer>
 #include <QPointer>
-#include <QTimer>
-#include <QTime>
+#include <QHash>
 #include <QMetaType>
 
+QT_BEGIN_NAMESPACE
+class QTimer;
+QT_END_NAMESPACE
+
 namespace GammaRay {
-class TimerInfo
+/*
+ */
+class TimerId
 {
+    friend uint qHash(const TimerId &);
+
 public:
     enum Type {
+        InvalidType,
+        // A QObject, no timer id
+        QQmlTimerType,
+        // A QObject, unstable timerId across stop/start
         QTimerType,
-        QObjectType,
-        QQmlTimerType
+        // A timerId attached to a QObject
+        QObjectType
     };
 
-    struct TimeoutEvent
-    {
-        QTime timeStamp;
-        int executionTime;
-    };
+    TimerId();
+    explicit TimerId(QObject *timer);
+    explicit TimerId(int timerId);
+
+    bool isValid() const;
 
-    explicit TimerInfo(QObject *timer);
-    explicit TimerInfo(int timerId);
     Type type() const;
-    void addEvent(const TimeoutEvent &timeoutEvent);
-    void setLastReceiver(QObject *receiver);
-    int numEvents() const;
+    quintptr address() const;
     QTimer *timer() const;
     QObject *timerObject() const;
     int timerId() const;
-    FunctionCallTimer *functionCallTimer();
-    QString wakeupsPerSec() const;
-    QString timePerWakeup() const;
-    QString maxWakeupTime() const;
-    int totalWakeups() const;
-    QString state() const;
-    QString displayName() const;
+
+    bool operator==(const TimerId &other) const;
+    bool operator==(QObject *timer) const;
+    bool operator==(int timerId) const;
+    bool operator<(const TimerId &other) const;
 
 private:
     Type m_type;
-    int m_totalWakeups;
-
-    // Only for QTimer/QQmlTimers timers
+    quintptr m_timerAddress;
     QPointer<QObject> m_timer;
-
     int m_timerId;
-    FunctionCallTimer m_functionCallTimer;
-    QList<TimeoutEvent> m_timeoutEvents;
+};
+
+struct TimerIdInfo
+{
+    TimerIdInfo()
+        : timerId(-1)
+        , totalWakeups(0)
+    { }
 
-    // Only for free timers, QObject that received the timeout event
-    QPointer<QObject> m_lastReceiver;
+    TimerIdInfo(const TimerId &id, QObject *receiver = nullptr);
+    virtual ~TimerIdInfo() { }
 
-    void removeOldEvents();
+    virtual void update(const TimerId &id, QObject *receiver = nullptr);
+
+    int timerId;
+    int totalWakeups;
+
+    QString objectName;
+    QString state;
```

Comment:
Does this (and the following 3) really need to be strings? Looking at moving tr() calls to the client, it would be better to have this as numeric values here, and only do the string conversion in the model on demand I think.